rfcs/0107-version-normalization.md

```diff
@@ -0,0 +1,220 @@
+nix-repl> builtins.parseDrvName "mpv-unstable-2021-05-03"
+builtins.parseDrvName "mpv-unstable-2021-05-03"
+{ name = "mpv-unstable"; version = "2021-05-03"; }
+```
+
+It happens because the `version` attribute in the set returned by
+`builtins.parseDrvName` always starts with a digit. This is a deliberate design
+decision of Nix specification, and as such it should not be regarded as a "bug";
+therefore, we should strive to follow it, neither circumventing nor ignoring it.
+
+Further, the `version` attribute should be crafted to satisfy the expected
+upgrading semantics stated in the manual pages, as effectively implemented by
+`builtins.compareVersions` -- even when the raw version of the original program
+does not meet this expectation.
+
+This document describes a format suitable to fix these issues, while keeping it
+understandable and striving for simplicity.
+
+# Detailed design
+[design]: #detailed-design
+
+## Terminology
+
+_Disclaimer_: whereas some of the terms enumerated below are borrowed from
+[pkgsrc guide](https://www.netbsd.org/docs/pkgsrc/) plus a bit of terminology
+employed by git, this document aims to be general, not being overly attached to
+them.
+
+- Program denotes the piece of software to be fetched and processed via Nixpkgs.
+  - This term makes no distinction about the immediate or intended uses of the
+    program; it can range from non-interactive programs to full GUI
+    applications, even including filesets intended as input to other programs
+    (such as firmwares and fonts).
+- Team denotes the maintainers of the program.
+  - This term makes no distinction among the various models of organization of a
+    team; it ranges from a solitary programmer to a business company - and
+    everything inbetween.
+- Source denotes the origin of the program.
+  - This term makes no distinction among precompiled, binary-only or high-level
+    code distributions.
+- Snapshot denotes the source of the program taken at a point of the time.
+  - Labeled snapshot denotes any snapshot explicitly labeled by its team.
+  - Unlabeled snapshot denotes any snapshot not explicitly labeled by its team.
+- Release denotes any distributed snapshot, as defined by its team.
+- Branch denotes a logical sequence of snapshots, as identified by the program's
+  team;
+  - Usually these branches are denoted by names such as `stable`, `master`,
+    `unstable`, `trunk`, `experimental`, `staging`, `X.Y` (where `X` and
+    `Y` are numbers) etc.
+
+## Design
+
+- For a labeled snapshot:
+  - `version` should be constituted of the version of the snapshot, as defined
+    by the program team, without any alphabetical characters (e.g. "v", "rel")
+    prepending it.
+    - Alphabetical characters following the first numerical character of version
+      (as defined above) should be maintained, except optionally those clearly
+      used as separators, in which case they are replaced by dots (emulating a
+      typical dot-separated version).
```

Comment:
Why allow replacing them?

Reply:
Because it is ugly to use `RELEASE_17_0_RC1` as version string.

Reply:
Ah, yes, that makes sense.  Although in my experience, that sort of naming is usually just used for git/svn tags, not how upstream actually refers to their releases.

Reply:
@AndersonTorres Can you add an example of such replacing?